src/ring_theory/witt_vector/compare.lean

```diff
@@ -0,0 +1,199 @@
+
+end iso
+
+end truncated_witt_vector
+
+namespace witt_vector
+open truncated_witt_vector
+
+variables {p : ℕ} [hp : fact p.prime]
+include hp
+
+local notation `𝕎` := witt_vector p -- type as `\bbW`
+
+variables (p)
+
+/--
+`to_zmod_pow` is a family of compatible ring homs. We get this family by composing
+`truncated_witt_vector.zmod_equiv_trunc` (in right-to-left direction)
+with `witt_vector.truncate`.
+-/
+def to_zmod_pow (k : ℕ) : 𝕎 (zmod p) →+* zmod (p ^ k) :=
+(zmod_equiv_trunc p k).symm.to_ring_hom.comp (truncate k)
+
+lemma to_zmod_pow_compat (m n : ℕ) (h : m ≤ n) :
+  (zmod.cast_hom (pow_dvd_pow p h) (zmod (p ^ m))).comp (to_zmod_pow p n) = to_zmod_pow p m :=
+calc (zmod.cast_hom _ (zmod (p ^ m))).comp
+      ((zmod_equiv_trunc p n).symm.to_ring_hom.comp (truncate n)) =
+  ((zmod_equiv_trunc p m).symm.to_ring_hom.comp
+    (truncated_witt_vector.truncate h)).comp (truncate n) :
+  by rw [commutes_symm, ring_hom.comp_assoc]
+... = (zmod_equiv_trunc p m).symm.to_ring_hom.comp (truncate m) :
+  by rw [ring_hom.comp_assoc, truncate_comp_witt_vector_truncate]
+
+/--
+`to_padic_int` lifts `to_zmod_pow` to a ring hom to `ℤ_[p]` using `padic_int.lift`, the universal
+property of `ℤ_[p]`.
+-/
+def to_padic_int : 𝕎 (zmod p) →+* ℤ_[p] := padic_int.lift $ to_zmod_pow_compat p
+
+lemma zmod_equiv_trunc_compat (k₁ k₂ : ℕ) (hk : k₁ ≤ k₂) :
+    (truncated_witt_vector.truncate hk).comp
+        ((zmod_equiv_trunc p k₂).to_ring_hom.comp
+           (padic_int.to_zmod_pow k₂)) =
+      (zmod_equiv_trunc p k₁).to_ring_hom.comp (padic_int.to_zmod_pow k₁) :=
+by rw [← ring_hom.comp_assoc, commutes, ring_hom.comp_assoc, padic_int.zmod_cast_comp_to_zmod_pow]
+
+/--
+`from_padic_int` uses `witt_vector.lift` to lift `truncated_witt_vector.zmod_equiv_trunc`
+composed with `padic_int.to_zmod_pow` to a ring hom `ℤ_[p] →+* 𝕎 (zmod p)`.
+-/
+def from_padic_int : ℤ_[p] →+* 𝕎 (zmod p) :=
+witt_vector.lift (λ k, (zmod_equiv_trunc p k).to_ring_hom.comp (padic_int.to_zmod_pow k)) $
+  zmod_equiv_trunc_compat _
+
+lemma to_padic_int_comp_from_padic_int :
+  (to_padic_int p).comp (from_padic_int p) = ring_hom.id ℤ_[p] :=
+begin
+  rw ← padic_int.to_zmod_pow_eq_iff_ext,
+  intro n,
+  rw [← ring_hom.comp_assoc, to_padic_int, padic_int.lift_spec],
+  simp only [from_padic_int, to_zmod_pow, ring_hom.comp_id],
+  rw [ring_hom.comp_assoc, truncate_comp_lift, ← ring_hom.comp_assoc],
+  convert ring_hom.id_comp _,
+end
+
+lemma to_padic_int_comp_from_padic_int_ext (x) :
+  (to_padic_int p).comp (from_padic_int p) x = ring_hom.id ℤ_[p] x :=
+by rw to_padic_int_comp_from_padic_int
+
+lemma from_padic_int_comp_to_padic_int :
+  (from_padic_int p).comp (to_padic_int p) = ring_hom.id (𝕎 (zmod p)) :=
+begin
+  apply witt_vector.hom_ext,
+  intro n,
+  rw [from_padic_int, ← ring_hom.comp_assoc, truncate_comp_lift, ring_hom.comp_assoc],
+  simp only [to_padic_int, to_zmod_pow, ring_hom.comp_id],
+  rw [padic_int.lift_spec, ← ring_hom.comp_assoc],
+  convert ring_hom.id_comp _,
```

Comment:
Same comment (with a variation on the `simp` lemma).

Reply:
Done.